src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3152,6 +3153,312 @@ void MacroAssembler::check_klass_subtype(Register sub_klass,
+
+  // Is there another entry to check? Consult the bitmap.
+  testbit(r_bitmap, (bit + 1) & Klass::SECONDARY_SUPERS_TABLE_MASK);
+  z_bfalse(L_failure);
+
+  // Linear probe. Rotate the bitmap so that the next bit to test is
+  // in Bit 2 for the look-ahead check in the slow path.
+  if (bit != 0) {
+    z_rllg(r_bitmap, r_bitmap, 64-bit); // rotate right
+  }
+
+  // Calls into the stub generated by lookup_secondary_supers_table_slow_path.
+  // Arguments: r_super_klass, r_array_base, r_array_index, r_bitmap.
+  // Kills: r_array_length.
+  // Returns: r_result
+
+  call_stub(StubRoutines::lookup_secondary_supers_table_slow_path_stub());
+
+  z_bru(L_done); // pass whatever result we got from a slow path
+
+  bind(L_failure);
+  // TODO: use load immediate on condition and z_bru above will not be required
+  z_lghi(r_result, 1);
+
+  bind(L_done);
+  BLOCK_COMMENT("} lookup_secondary_supers_table");
+
+  if (VerifySecondarySupers) {
+    verify_secondary_supers_table(r_sub_klass, r_super_klass, r_result,
+                                  r_temp1, r_temp2, r_temp3);
+  }
+}
+
+// Called by code generated by check_klass_subtype_slow_path
+// above. This is called when there is a collision in the hashed
+// lookup in the secondary supers array.
+void MacroAssembler::lookup_secondary_supers_table_slow_path(Register r_super_klass,
+                                                             Register r_array_base,
+                                                             Register r_array_index,
+                                                             Register r_bitmap,
+                                                             Register r_result,
+                                                             Register r_temp1) {
+  assert_different_registers(r_super_klass, r_array_base, r_array_index, r_bitmap, r_result, r_temp1);
+
+  const Register
+    r_array_length = r_temp1,
+    r_sub_klass    = noreg;
+
+  LOOKUP_SECONDARY_SUPERS_TABLE_REGISTERS;
+
+  BLOCK_COMMENT("lookup_secondary_supers_table_slow_path {");
+  NearLabel L_done, L_failure;
+
+  // Load the array length.
+  z_llgf(r_array_length, Address(r_array_base, Array<Klass*>::length_offset_in_bytes()));
+
+  // And adjust the array base to point to the data.
+  // NB!
+  // Effectively increments the current slot index by 1.
+  assert(Array<Klass*>::base_offset_in_bytes() == wordSize, "");
+  add2reg(r_array_base, Array<Klass*>::base_offset_in_bytes());
+
+  // Linear probe
+  NearLabel L_huge;
+
+  // The bitmap is full to bursting.
+  // Implicit invariant: BITMAP_FULL implies (length > 0)
+  assert(Klass::SECONDARY_SUPERS_BITMAP_FULL == ~uintx(0), "");
+
+  z_cghi(r_bitmap, Klass::SECONDARY_SUPERS_BITMAP_FULL);
+  z_bre(L_huge);
+
+  // NOTE: please load 0 only in r_result, as this is also being used for z_locgr down
+  clear_reg(r_result, true /* whole_reg */, false /* set_cc */); // let's hope that search will be a success
```

Comment:
Is this needed? It should be already done by `lookup_secondary_supers_table`. Could be asm_assert'ed.